### ISSUE TITLE
Mark nodes as "NotReady" until critical pods from daemonsets are Ready

### ISSUE THREAD

**Issue by lbernail** [kubernetes/kubernetes#75890]
**What would you like to be added**:
An extension to node conditions that would mark them NotReady until critical pods from daemonsets are running (maybe a CriticalPodsReady or DaemonsetsReady similar to the NetworkReady condition).

**Why is this needed**:
When using cluster-autoscaler, application pods are immediately scheduled when a node starts and sometimes they start before local pods from daemonsets are Ready, which can be an issue. A few examples:
- kube-proxy, which prevents application from connecting to their dependencies (DNS is the most common problem)
- local dns pod: we deploy a local dns cache as a daemonset and configure pods to use it as a resolver, so resolution fail until this pod is up
- monitoring agent: some applications send metrics very early when they start and require the monitoring agent to be ready
- kube2iam: required by some applications to get aws credentials

/sig scheduling
@kubernetes/sig-scheduling-feature-requests

**Comment by Joseph-Irving** [comment]
We very recently made a controller to deal with this exact issue. https://github.com/uswitch/nidhogg tainting the node if certain daemonset pods haven't been scheduled and are ready yet.

**Comment by lbernail** [comment]
> We very recently made a controller to deal with this exact issue. https://github.com/uswitch/nidhogg tainting the node if certain daemonset pods haven't been scheduled and are ready yet.

@Joseph-Irving Very nice. I'll definitely have a look. It would be great to have it natively in kubernetes but this would work in the meantime (and if the feature is not accepted of course)

**Comment by yastij** [comment]
cc @kubernetes/sig-scheduling-feature-requests @kubernetes/sig-node-feature-requests 

I think this is something worth exploring. This can be tied to priorityClass: mark nodes with a specific condition until pods of a specific priorityClass are running on the node.

**Comment by Huang-Wei** [comment]
I like the idea. We have [pod ready++](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md), why not node ready++?

**Comment by yastij** [comment]
/assign

I’ll brainstorm with @lbernail and see if can come up with a KEP to kickoff discussions.

**Comment by cizixs** [comment]
We have the same feature requirement, and have come with up a design and implementation running in our kubernetes cluster.

And we would love to share and contribute to this.

**Comment by yastij** [comment]
@cizixs - can you share some thoughts and use cases here ?

**Comment by kevin-wangzefeng** [comment]
> We very recently made a controller to deal with this exact issue. https://github.com/uswitch/nidhogg tainting the node if certain daemonset pods haven't been scheduled and are ready yet.

Having a controller to add taints to block node readiness during initialization may not work well under race condition. To complete this solution, you can make nodes born with that taint.

**Comment by cizixs** [comment]
> @cizixs - can you share some thoughts and use cases here ?

## User Case

* CNI service: we implemented CNI with a  running service, pods can only be scheduled to the node when CNI service is up and running
* monitor agent: monitoring data is very important for us, running application with no metrics is considered serious problem
* ...

## Design

Add a APIServer admission, add a taint to newly added node, so nodes are not schedule-able even if kubelet is running.

Define a CRD, and each node has a corresponding CR(with readinessGates field) indicates if the node is ready++. We use a CRD instead of adding a new node field, so node spec is keep unchanged, and it's easy to keep update with community.

Daemonset pods on the node is responsible for updating node CR readiness status that should concern them.

Add a new controller that creates node CR when a node is created(readinessGate list is read from a configmap), and also listens to node readiness CR to add or delete node taint based on if all readinessGate is ready.

**Comment by cizixs** [comment]
@yastij The above is the main logic, we can help to proposal a KEP to illustrate in detail.

**Comment by andrewsykim** [comment]
Initial node readiness gate KEP is here https://github.com/kubernetes/enhancements/pull/1003. I'll be iterating on it during the v1.15 cycle and hopefully we can have initial implementation in time for v1.16

**Comment by javierprovecho** [comment]
Another use case to the list: daemonset logging agents.

> https://github.com/kubernetes/kubernetes/issues/75890#issuecomment-485069650
> To complete this solution, you can make nodes born with that taint.

If nodes are born with that taint, pending pods won’t trigger the cluster-autoscaler because lack of toleration, so a scale up would never take place.

**Comment by lbernail** [comment]
/remove-lifecycle stale
@yastij or @andrewsykim I think this feature would have real value. Anyway I could help making it progress?

**Comment by aboyett** [comment]
Critical daemonsets are a necessary part of a secure, instrumented cluster. Whether it is a logging daemon, CNI-provider or the GKE metadata service it should be possible to prevent workloads from being scheduled on a node until these critical pods are running.

**Comment by integrii** [comment]
We're seeing instances in our clusters wherein `kube-proxy` being provisioned races with other containers being built.  This results in missing netfilter rules and services appear to be down for a short period of time.  We detect this blip sometimes with [Kuberhealthy](github.com/Comcast/kuberhealthy), which is our software that runs synthetic checks.  Our DNS checker often spins up very fast on new nodes and instantly fails to resolve DNS.

I can see that this would be challenging fix to implement given the existing scheduler logic:  How can the scheduler rapidly determine which pods must be on a node before others considering taints, selectors, regions, zones, resources, and everything else?

Perhaps the simplest way to do this _is_ to start nodes with a taint and implement a controller that runs on master nodes and removes taints appropriately.

**Comment by herikwebb** [comment]
We're seeing this issue as well. We run a machine learning platform and many of the workloads trigger significant scale-ups causing new nodes and new [kiam](https://github.com/uswitch/kiam) agents to be spun up and often times the applications are attempting to fetch aws credentials before the agent is ready resulting in failures.

**Comment by saberrey** [comment]
Hi, everyone! I also see this problem recently, I designed a method to sort the images according to the pod's priorityclassname and namespace when pulling the images, so that the critical pods' images can be pulled in the priority order to achieve the purpose of starting high-priority services first. Is this thought correct? [PR here: [https://github.com/kubernetes/kubernetes/pull/93239](url)]

However, there are still some problems in my implementation. In order to sort the pull requests in the channel, I need to wait a short period of time to allow as many requests as possible to enter the channel when taking the pull request, otherwise the first pull request that arrives will not be sorted and pull immediately. This strategy of waiting for a while is not elegant, does anyone have a better idea?

**ps: It only works under serial pull strategy( --serialize-image-pulls=true )**

**Comment by smarterclayton** [comment]
It would be better never to use serial pull strategy if at all possible,
since that is going to give the worst performance except when a registry is
already pathologically bad.

**Comment by jacksontj** [comment]
For anyone else having this issue, we ended up creating a controller to manage this. After reading through this thread and some other issues (e.g. https://github.com/kubernetes/enhancements/pull/1003#issuecomment-619087019) we decided that we'd go with the proposed solution -- namely nodes start with a taint which is then removed. We created https://github.com/wish/nodetaint to do so-- with this controller you can annotate certain daemonsets as "required" and only once that DS is running on the node will the taint be removed. We've been using this for a few months now to great success. It nicely fits into the k8s ecosystem as you can still launch pods before this controller has removed the taint (by using tolerations on the pod). So overall seems like an elegant solution to the problem that fits into the k8s ecosystem. I still think it would be nice to have some built-in way to do the same; but with this controller there is a solution out there that people can use while this issue works itself out.

**Comment by adisky** [comment]
This issue can be solved using the `taint` based approach as mentioned here
https://github.com/kubernetes/enhancements/pull/1003#issuecomment-619087019

and here
https://github.com/kubernetes/kubernetes/issues/75890#issuecomment-725792993
I think we can safely close this, the related KEP also has been closed with the same advice

/kind support
/close

**Comment by innovate-invent** [comment]
@adisky I don't think this issue is resolved. Given how common daemonsets are, k8s shouldn't need a third party tool to ensure that the correct scheduling order occurs. This is mandatory for CSI drivers to function properly.

**Comment by MatthewFagan** [comment]
The taints-based approach is broken when combined with cluster autoscaler. The autoscaler creates a new node and waits for it to be ready. When it tries to schedule the pending pod onto the new node, it sees the taints (because the daemonsets are not ready), and so it scales up again. It will keep scaling up for as long as it takes for the daemonsets to become ready. So it is not really a proper solution

**Comment by innovate-invent** [comment]
I think that is an issue with the autoscaler at that point. It should be able to be configured to understand temporary taints.

**Comment by sebltm** [comment]
I still think this would be helpful, although you can simulate the same thing with a controller and taints it just feels like more of a hack. We would like to perform some actions on the node before it's marked as Ready using a daemonset, but currently for an ASG node, we have to join the node (and therefore have some pods scheduled), get our controller to catch that new node, add a ``NoExecute`` taint to kick out the scheduled pods, perform our actions and then remove the taint...

**Comment by jacksontj** [comment]
FYI the k8s cluster autoscaler has a flag `ignore-taint` (added [here](https://github.com/kubernetes/autoscaler/pull/1963/files)) which enables you to use that `node-taint` mechanism (I have also updated the [README](https://github.com/wish/nodetaint) to include this information). I do agree this feels a bit hacky but these pieces are in-place we haven't been having issues with it -- so overall seems to practically work great.

**Comment by chenk008** [comment]
All controllers need to be compatible with the taints-based approach, I think it is brought extra work for component maintainers. A generic `ready` approach is needed.

**Comment by matglas** [comment]
The is a new alpha feature in scheduler. It allows pod to have scheduling paused when a controller set the schedulingGate.
https://github.com/kubernetes/kubernetes/pull/113274

This might be valuable for this.

**Comment by ChipmunkV** [comment]
So what's the solution?

For example for a cluster on GKE we want our custom DaemonSet in addition to the DaemonSets that GKE uses to configure the node (they installs GPU drivers). I imagine GKE uses some taints autoscaler is aware of. I don't understand how do we add another taint on top of that.

**Comment by balonik** [comment]
Anyone tried Pod Affinity for this? The doc says that it will schedule Pod on a node only if there is already another Pod with set labels running, but it doesn't specify what "running" means. If it must be in `Ready` state or other states are valid as well.

**Comment by johnko** [comment]
> Anyone tried inter Pod Affinity for this? The doc says that it will schedule Pod on a node only if there is already another Pod with set labels running, but it doesn't specify what "running" means. If it must be in `Ready` state or other states are valid as well.

just tried on 1.29: podAffinity seems like it doesn't work because it will still create the deployment pod on the node if the daemonset pod i set affinity to is in Error/CrashLoopBackOff status
